spec/unit/deep_preloader_spec.rb

```diff
@@ -0,0 +1,259 @@
+    end
+
+    failure_message do |actual|
+      assoc = actual.association(association_name)
+      if !assoc.loaded?
+        "expected #{actual.inspect} to have loaded #{association_name}"
+      else
+        "expected #{actual.inspect} to have loaded #{association_name} as #{@target.inspect}, but was #{assoc.target.inspect}"
+      end
+    end
+  end
+
+  it "can parse a hash spec" do
+    spec          = { a: [:b, { c: :d, e: nil }]}
+    expected_spec = DeepPreloader::Spec.new(a: DeepPreloader::Spec.new(b: nil, c: DeepPreloader::Spec.new(d: nil), e: nil))
+
+    expect(DeepPreloader::Spec.parse(spec)).to eq(expected_spec)
+  end
+
+  context "with a test model" do
+    with_temporary_table(:model){}
+
+    it "can parse a polymorphic hash spec" do
+      parsed_spec   = DeepPreloader::Spec.parse(a: DeepPreloader::PolymorphicSpec.parse("Model" => { b: :c }))
+      expected_spec = DeepPreloader::Spec.new(a: DeepPreloader::PolymorphicSpec.new("Model" => DeepPreloader::Spec.new(b: DeepPreloader::Spec.new(c: nil))))
+
+      expect(parsed_spec).to eq(expected_spec)
+    end
+  end
+
+  context "with a one-to-one relationship" do
+    with_temporary_table(:parent, ->(t){ t.references :child }) do
+      belongs_to :child, inverse_of: :parent
+    end
+
+    with_temporary_table(:child) do
+      has_one :parent, inverse_of: :child
+    end
+
+    let!(:child) { Child.create! }
+    let!(:parent) { Parent.create!(child: child).tap(&:reload) }
+    before(:each) { child.reload }
+
+    context "in the belongs_to direction" do
+      let!(:childless) { Parent.create! }
+
+      it "is not already loaded" do
+        expect(parent.association(:child)).to_not be_loaded
+      end
+
+      it "loads the child" do
+        DeepPreloader.preload(parent, :child)
+        expect(parent).to have_loaded(:child).as(child)
+      end
+
+      it "sets up the inverse relationship" do
+        DeepPreloader.preload(parent, :child)
+        expect(parent.child).to have_loaded(:parent).as(parent)
+      end
+
+      it "satisfies a childless parent" do
+        DeepPreloader.preload(childless, :child)
+        expect(childless).to have_loaded(:child).as(nil)
+      end
+
+      it "loads more than one entity" do
+        DeepPreloader.preload([parent, childless], :child)
+        [parent, childless].zip([child, nil]).each do |p, c|
+          expect(p).to have_loaded(:child).as(c)
+        end
+      end
+
+      it "uses preloaded children from other parents before hitting the database" do
+        loaded_parent = Parent.create!(child: child)
+        Child.where(id: child).delete_all
+
+        DeepPreloader.preload([parent, loaded_parent], :child)
+        [parent, loaded_parent].each do |p|
+          expect(p).to have_loaded(:child).as(child)
+        end
+      end
+
+      it "supports locking" do
+        DeepPreloader.preload(parent, :child, lock: "FOR SHARE")
+        # No real way to test this with sqlite: AR drops the lock clause.
```

Comment:
The difficulty here is not integrating postgres, but making good use of that. If we want postgres, making the spec_helper setup react to environment variables seems reasonable. Currently though, not useful.